tabbycat/notifications/consumers.py

```diff
@@ -17,9 +18,16 @@
                     motion_release_email_generator, randomized_url_email_generator,
                     standings_email_generator, team_draw_email_generator, team_speaker_email_generator)
 
+if TYPE_CHECKING:
+    from django.db.models import QuerySet, Model
+
 
 class NotificationQueueConsumer(SyncConsumer):
 
+    NOTIFICATION_GENERATORS: Dict[str, Union[
+        Callable[['QuerySet[Person]', str, 'Model'], List[Tuple[Dict[str, str], Person]]],
+        Callable[['QuerySet[Person]', 'Model'], List[Tuple[Dict[str, str], Person]]],
+    ]]
```

Comment:
Ugly typing, but also `BulkNotification.EVENT_TYPE_` should also be an `enum` (like email status types).

Reply:
Did a bit of research; a `Protocol` might be in order for this:
```py
from typing import Any, List, Protocol, Tuple

class NotificationGenerator(Protocol):
    def __call__(self, *vals: Any) -> List[Tuple[EmailContextData, Person]]: 
```

Then, the type annotation for `NOTIFICATION_GENERATORS` could be (including my other suggestions):

```py
NOTIFICATION_GENERATORS: Dict[BulkNotification.EventType, NotificationGenerator]
```
which is much clearer yet flexible.